app/scripts/background.js

```diff
@@ -405,14 +412,47 @@ export async function loadStateFromPersistence() {
   versionedData =
     (await localStore.get()) || migrator.generateInitialState(firstTimeState);
 
+  const lastGoodMigration = global.localStorage.getItem('lastGoodMigration');
+  const lastGoodMigrationNum =
+    typeof Number(lastGoodMigration) === 'number'
+      ? Number(lastGoodMigration)
+      : 0;
+  const currentVersionNumber = versionedData.meta?.version;
+
+  if (lastGoodMigration && lastGoodMigrationNum !== currentVersionNumber) {
+    statePersistenceEvents.emit('version-mismatch', {
+      vaultStructure: getObjStructure(versionedData),
+    });
+  }
+
   // check if somehow state is empty
   // this should never happen but new error reporting suggests that it has
   // for a small number of users
   // https://github.com/metamask/metamask-extension/issues/3919
   if (versionedData && !versionedData.data) {
     // unable to recover, clear state
     versionedData = migrator.generateInitialState(firstTimeState);
-    sentry.captureMessage('MetaMask - Empty vault found - unable to recover');
+    const keyringVault = global.localStorage.getItem('metaMaskVault');
+    if (keyringVault) {
+      console.log(versionedData);
+      versionedData.data.KeyringController = {
+        // Restore the vault from localstorage
+        vault: JSON.parse(keyringVault),
+        // Set a flag to indicate that the vault was restored from backup
+        restoredFromBackup: true,
+      };
+      // We have to set the completedOnboarding flag to true to avoid an
+      // Infinite routing loop where after unlocking they are redirected back
+      // to an unlock screen on the onboarding flow.
+      versionedData.data.OnboardingController = {
+        completedOnboarding: true,
+      };
+    }
+    // sentry.captureMessage('MetaMask - Empty vault found - unable to recover');
```

Comment:
Me commenting out this line was required before, and obfuscated my errors with the misMatch of lastGoodVersion Number. will clean this up soon.